### PR TITLE
Add gan_system.py

### PR DESCRIPTION
**Pull request by JorisCos**
This PR separates gan_system.py from the SEGAN recipe.
The code has been changed to provide a more general approach of GAN's training

### CODE REVIEW

File: asteroid/engine/gan_system.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+from collections import OrderedDict
+import soundfile as sf
+import pytorch_lightning as pl
+import torch
+import os
+
+
+class GanSystem(pl.LightningModule):
+
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:9]: Need docstring

File: asteroid/engine/gan_system.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    def __init__(self, discriminator, generator, opt_d, opt_g,
+                 scheduler_d, scheduler_g, discriminator_loss, generator_loss,
+                 validation_loss, train_loader, val_loader=None, conf=None):
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:12]: I don't think the schedulers should be positional arguments (meaning mandatory), they should be keyword args with None as default. 
Is there a good default value for validation loss or is it mandatory? I think we can make it a kwarg as well.

File: asteroid/engine/gan_system.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    def on_batch_end(self):
+        """ Overwrite if needed. Called by pytorch-lightning"""
+        pass
+
+    def on_epoch_start(self):
+        """ Overwrite if needed. Called by pytorch-lightning"""
+        pass
+
+
+    @pl.data_loader
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:155]: The decorator is not needed anymore.

File: asteroid/engine/gan_system.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.scheduler_d = scheduler_d
+        self.scheduler_g = scheduler_g
+        self.conf = conf
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def routine_step(self, batch, optimizer_idx):
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:31]: If routine_step is the same as training_step, what's the goal of routine step?

Comment (mpariente) [reply asteroid/engine/gan_system.py:31]: I wonder if it wouldn't be beneficial to separate `generator_training_step`, `discriminator_training_step` or it's not generic enough?

File: asteroid/engine/gan_system.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        self.g_loss = generator_loss
+        self.validation_loss = validation_loss
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.scheduler_d = scheduler_d
+        self.scheduler_g = scheduler_g
+        self.conf = conf
+
+    def forward(self, z):
+        return self.generator(z)
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:29]: I would have *args, **kwargs here. It would be more generic for conditional GANs for example

File: asteroid/engine/gan_system.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        if optimizer_idx == 0:
+            # Compute D loss for targets (labels = 1)
+            est_true_labels = self.discriminator(targets, inputs, targets)
+            true_loss = self.d_loss(inputs, targets, estimates,
+                                    est_true_labels, true_labels)
+            # Compute D loss for estimates (labels = 0)
+            est_false_labels = self.discriminator(estimates.detach(), inputs,
+                                                  targets)
+            fake_loss = self.d_loss(inputs, targets, estimates,
+                                    est_false_labels, false_labels)
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:51]: I don't like that many positional arguments, it's very error-prone. Two or three would be a maximum, five is too much for me.

File: asteroid/engine/gan_system.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            # Overall, the loss is the mean of these
+            d_loss = (true_loss + fake_loss) * 0.5
+            tqdm_dict = {'d_loss': d_loss}
+            output = OrderedDict({
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:55]: Curiosity, why is it ordered?

File: asteroid/engine/gan_system.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+    def routine_step(self, batch, optimizer_idx):
+
+        # Get data from data_loader
+        inputs, targets = batch
+        # Forward inputs
+        estimates = self(inputs)
+        # The discriminator is basically a binary classifier
+        true_labels = True
+        false_labels = False
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:39]: Why defining them here and not using True and False? For soft labels? Something else I don't see?

File: asteroid/engine/gan_system.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            batch: the object returned by the loader (a list of torch.Tensor
+                in most cases) but can be something else.
+            batch_nb (int): The number of the batch in the epoch.
+
+        Returns:
+            dict:
+
+            ``'val_loss'``: loss
+        """
+        inputs, targets = batch
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:93]: By default, don't we want d_loss and g_loss at validation time? 
I'm no GAN expert, just asking.

File: asteroid/engine/gan_system.py (hunk 2 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:12]: I guess stoi is the best choice.

File: asteroid/engine/gan_system.py (hunk 4 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:31]: This is a legacy from the previous version and should be in the training step.
I find it simpler to have all the training in the training step but it doesn't cost much to create `generator_training_step` `discriminator_training_step` as you wish.

File: asteroid/engine/gan_system.py (hunk 5 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:29]: I don't understand how we can pass more arguments to the generator anyway since the call is done in `gan_system` and is never to be changed by the user or am I missing something ?

File: asteroid/engine/gan_system.py (hunk 7 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:55]: This was taken from the PL docs,

File: asteroid/engine/gan_system.py (hunk 8 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:39]: No you are right this useful when don't separate the generator loss and the discriminator loss but we do.

File: asteroid/engine/gan_system.py (hunk 9 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:93]: In the validation step we only pass data through the generator and evaluate its performance.
The value of d_loss and g_loss is not very relevant at this point because

File: asteroid/engine/gan_system.py (hunk 6 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:51]: I understand your point but I don't know how I can do a generic call to d_loss with less positional args it's the same problem as my previous comment

File: tests/engine/gan_system_test.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        super().__init__()
+
+    def forward(self, estimates, targets, est_labels):
+        loss = torch.mean((est_labels - torch.ones_like(est_labels)) ** 2)
+        return loss
+
+
+class DiscriminatorLoss(_Loss):
+    def __init__(self):
+        super().__init__()
```
Comment (mpariente) [inline tests/engine/gan_system_test.py:34]: Those lines are not needed. 
If you don't subclass a method, it stays the same. 
Same above

File: asteroid/engine/gan_system.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        self.estimates = None
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def training_step(self, batch, batch_nb, optimizer_idx):
+        # Get data from data_loader
+        inputs, targets = batch
+        # Forward inputs
+        self.estimates = self(inputs)
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:65]: Why do you need `estimates` as an attribute?

Comment (mpariente) [reply asteroid/engine/gan_system.py:65]: Also, I don't know how well this will behave when we train on multi GPU.

File: asteroid/engine/gan_system.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        # Train discriminator
+        if optimizer_idx == 0:
+            # Compute D loss for targets
+            est_true_labels = self.discriminator(targets, inputs, targets)
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:69]: Is this discriminator call completely generic?

**Comment by mpariente** [comment]
@popcornell Can you have a a look please?

File: asteroid/engine/gan_system.py (hunk 9 above)
Comment (mpariente) [reply asteroid/engine/gan_system.py:113]: Because what?

File: asteroid/engine/gan_system.py (hunk 5 above)
Comment (mpariente) [reply asteroid/engine/gan_system.py:59]: Different GANs will have different `nn.Module` as generator. 
```suggestion
    def forward(self, *args, **kwargs):
        return self.generator(*args, **kwargs)
```
This is just more generic.

File: asteroid/engine/gan_system.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+               of learning rate schedulers for d.
+           scheduler_g (torch.optim.lr_scheduler._LRScheduler): Instance
+               of learning rate schedulers for g.
+           conf: Anything to be saved with the checkpoints during training.
+               The config dictionary to re-instantiate the run for example.
+       """
+
+    def __init__(self, discriminator, generator, opt_d, opt_g,
+                 discriminator_loss, generator_loss, train_loader,
+                 validation_loss=PITLossWrapper(NegSTOILoss(), pit_from='pw_pt'),
```
Comment (mpariente) [inline asteroid/engine/gan_system.py:39]: I would put `None` as the validation loss, and we don't run validation step if we don't have validation loss. 
Imposing STOI doesn't seem right to me.

File: asteroid/engine/gan_system.py (hunk 9 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:113]: Ahah sry. Because they don't give information about the performance, and they will converge to a value after few iterations, so I don't think it's worth the computation

File: asteroid/engine/gan_system.py (hunk 11 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:65]: This was done in case someone wants to use estimates during  on_epoch_ends or on_training_end etc to save examples or see the evolution of one estimates.

File: asteroid/engine/gan_system.py (hunk 12 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:69]: I think so, we basically give everything to the network in case someone need the true ref or the noisy ref

Comment (mpariente) [reply asteroid/engine/gan_system.py:69]: But this is written nowhere, so the user has to come here to find what the `GanSystem` expects of his network. If he comes this far to read the code, he can re-implement his own System. 
I'm not sure this will be useful. It will either require huge amount of docs or be too simplistic. 
And the code is not so because most args are positional.. 
Maybe it's useful to have the backbone with the optimizer, the scheduler etc.. so that it doesn't have to be rewritten. However, the training step should either be super simple, or `NotImplementedError` IMO. @popcornell any thoughts on that? You know GAN way better than me.

File: asteroid/engine/gan_system.py (hunk 11 above)
Comment (popcornell) [reply asteroid/engine/gan_system.py:65]: but wouldn't be better to save only validation examples ?

File: asteroid/engine/gan_system.py (hunk 12 above)
Comment (popcornell) [reply asteroid/engine/gan_system.py:69]: why it does take 3 arguments with two arguments identical ? Plain discriminator should have one argument (one time is the enhanced signal another time is the true clean signal). Are you using relativistic discriminator ?

Comment (popcornell) [reply asteroid/engine/gan_system.py:69]: > Maybe it's useful to have the backbone with the optimizer, the scheduler etc.. so that it doesn't have to be rewritten. However, the training step should either be super simple, or NotImplementedError IMO. @popcornell any thoughts on that? You know GAN way better than me.

GANs are quite sensitive and difficult to train.... plus there are tons of different types some differs only for the output activation of discriminator and loss function but others (like relativistic discriminator ) requires changes to the training loop.  Also because are hard to train one usually logs everything to tensorboard in the training loop. I mean it is nice to have a GAN example and maybe a GAN system. But for real research it will be bypassed probably.

Comment (popcornell) [reply asteroid/engine/gan_system.py:69]: I will actually like to do some barebone system which bypasses most of lightning functionality (also grad accumulation). For example one could wish to have different virtual batches for generator and discriminator ( now with wasserstein gan and such is less needed but still...) . Also useful for other techniques which require full control over the training loop.

Comment (mpariente) [reply asteroid/engine/gan_system.py:69]: Thanks for having a look !

**Comment by mpariente** [comment]
Well IDK @JorisCos, maybe have the "complicated" (hear varying from one arch to another) steps stay as `NotImplementedError` so that it's clear we don't take care of that?

File: asteroid/engine/gan_system.py (hunk 11 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:65]: We don't make it mandatory to have a validation step, but we always have a training step.
`GanSystem `could take an arg that specifies if we want to save examples `on_training_end` or `on_epoch_end` but we don't do it in `System `I think I will just remove the attribute.

File: asteroid/engine/gan_system.py (hunk 12 above)
Comment (JorisCos) [reply asteroid/engine/gan_system.py:69]: > why it does take 3 arguments with two arguments identical ? Plain discriminator should have one argument (one time is the enhanced signal another time is the true clean signal). Are you using relativistic discriminator ?

Sometimes the discriminator requires more than just the enhanced signal or the true clean signal. Some discriminator actually compare the enhanced signal to the true Signal ( ex MetricGAN)  some discriminator use the enhanced or the clean signal + the noisy input (SEGAN). This is why I made the discriminator have the args.

**Comment by JorisCos** [comment]
> Well IDK @JorisCos, maybe have the "complicated" (hear varying from one arch to another) steps stay as `NotImplementedError` so that it's clear we don't take care of that?

Just to be sure we agree on what to do next. I will go back to the first version of `GanSystem`  that had a `routine_step` which is specific to each GAN and that implements the training loop ? Which mean that for each new GAN based recipe, the  `routine_step` will have do be edited.

File: asteroid/engine/gan_system.py (hunk 12 above)
Comment (mpariente) [reply asteroid/engine/gan_system.py:69]: For the current use cases, this choice can make sense. But the audio GAN landscape is so broad that we cannot aim at covering it all..

**Comment by mpariente** [comment]
I'd stick with:
```python
def training_step(etc, ...):
    raise NotImplementedError
```

**Comment by mpariente** [comment]
Now that this is used in both SEGAN and metricGAN, I think we can merge this. 
Then, we'll get back to MetricGAN and SEGAN.
